cmd/api/insert.go

```diff
@@ -54,43 +45,71 @@ func (a *API) Insert(c *fiber.Ctx) error {
 			Msg("Incoming request")
 	}
 
+	body := c.Body()
+	if !gjson.ValidBytes(body) {
+		return fiber.NewError(http.StatusBadRequest, "invalid JSON")
+	}
+
 	// TODO: this block can be abstracted as we also use it for query
-	/////
 	apiKey := c.Locals("apiKey").(models.APIKey)
 
 	// TODO: read-only vs read-write connections
 	connectionSetting := a.db.GetDatabaseConnection(apiKey.DestinationID)
-
 	if connectionSetting.ID == "" {
-		return errors.New("No DB Connections set up")
+		return fiber.NewError(http.StatusUnauthorized, "no connection is set up")
 	}
-	/////
 
-	input := c.Body()
-	if !gjson.ValidBytes(input) {
-		return errors.New("Invalid JSON")
+	var (
+		tableName, flatAlgo string
+		fromBody            bool
+	)
+	flatAlgo = c.Get(FlattenHeader, c.Query(FlattenQuery))
+	tableName = c.Get(TableNameHeader, c.Query(TableNameQuery))
+	if tableName == "" {
+		if tableName = gjson.GetBytes(body, TableNameJson).String(); tableName == "" {
+			return fiber.NewError(http.StatusBadRequest, "missing required table field")
+		}
+		fromBody = true
 	}
 
-	// TODO: Use actual table name from request
-	tableName := "t"
-
-	var FlattenFunc = Flatten
```

Comment:
My intention here was to have a sort of generic `Flatten` interface, where we could have different implementations. This way the logic in the API call could be separate from how we flatten the JSON. Is there a way we can have a better separation?

Reply:
i think a function that delegates to the appropriate implementation should be enough. Something like https://github.com/sd1li/ScratchDB/commit/a9fc7c092452684a0a75f744cae81612204da302#diff-7c8ea8145bc2dc2f03d63f005eee621fc8825e621e4e5a2eefc16c7369cda451R123